### PR TITLE
[PAY-920] K2Q, K2r, and K2RQ factors

### PR DESCRIPTION
**Pull request by mattravenhill-humi**
https://gethumi.atlassian.net/browse/PAY-920

### CODE REVIEW

File: lib/taxman/taxman2023/calculate.rb [hunk 1]
```diff
@@ -40,6 +40,10 @@ def call
 
       context[:a] = A.amount(context)
 
+      context[:k2q] = K2Q.amount(context)
+      context[:k2q] = K2R.amount(context)
```
Comment (mark-h-humi) [inline lib/taxman/taxman2023/calculate.rb:44]:
```suggestion
      context[:k2r] = K2R.amount(context)
```

File: lib/taxman/taxman2023/k2rq.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+module Taxman2023
+  # Calculates the k2p factor for Quebec
+  class K2RQ < Factor
```
Comment (mark-h-humi) [inline lib/taxman/taxman2023/k2rq.rb:5]: same re filename/classname

File: lib/taxman/taxman2023/k2q.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+module Taxman2023
+  # Calculates the k2p factor for Quebec
+  class K2Q < Factor
```
Comment (mark-h-humi) [inline lib/taxman/taxman2023/k2q.rb:5]: same re filename/classname

File: lib/taxman/taxman2023/k2r.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    end
+
+    def maxium_pension_rate
+      3123.45
+    end
+
+    def maximum_ei
+      1002.45
+    end
+  end
```
Comment (mark-h-humi) [inline lib/taxman/taxman2023/k2r.rb:37]: constants?

File: lib/taxman/taxman2023/k2rq.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      3407.40
+    end
+
+    def maximum_ei
+      781.05
+    end
+
+    def maximum_insurable_earnings
+      449.54
+    end
```
Comment (mark-h-humi) [inline lib/taxman/taxman2023/k2rq.rb:48]: constants

File: lib/taxman/taxman2023/k2r.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      ].min * 0.15
+    end
+
+    def maxium_pension_rate
+      3123.45
+    end
+
+    def maximum_ei
+      1002.45
+    end
```
Comment (claus-humi) [inline lib/taxman/taxman2023/k2r.rb:36]: We can find this value in `Taxman2023::Ei::EI_MAX`. That max is for T4127 so we will need to add two another value in there for QC.

**Comment by mark-h-humi** [comment]
also specs please and thank-you @mattravenhill-humi

File: lib/taxman/taxman2023/k2r.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    attr_reader(*params)
+
+    def amount
+      [pension_amount + employment_insurance_amount]
+    end
+
+    def pension_amount
+      [
+        p * c * (0.0495 / 0.0595),
+        (maxium_pension_rate * (pm / 12)) - (dq * (0.0495 / 0.0640)) + (dq * (0.0540 / 0.0640))
```
Comment (claus-humi) [inline lib/taxman/taxman2023/k2r.rb:19]: I believe some of these constants are the rates of CPP and QPP.

**Review by claus-humi — COMMENTED** [review]
Generally looks good. We need to use the constants from other classes and we need specs for these classes with some sanity checks until we get scenarios from Greg that includes actual numbers for the whole moving in and out of Quebec thing. This will be a pain to debug in the future since we don't have a reference for these values.

File: lib/taxman/taxman2023/k2r.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module Taxman2023
+  # Calculates the k2p factor for Quebec
+  class K2R < Factor
```
[the remark this answers is not in the document]
Comment (mattravenhill-humi) [reply lib/taxman/taxman2023/k2r.rb]: All relevant file names have been updated to be camel-cased

File: lib/taxman/taxman2023/k2q.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+      449.54
+    end
+
+    def lower_qpp_rate
+      0.0540
+    end
+
+    def higher_qpp_rate
+      0.0640
+    end
```
[the remark this answers is not in the document]
Comment (mattravenhill-humi) [reply lib/taxman/taxman2023/k2q.rb]: These helper methods now reference constants and live in lib/taxman/taxman2023/k2_generic.rb

File: lib/taxman/taxman2023/k2r.rb (hunk 6 above)
Comment (mattravenhill-humi) [reply lib/taxman/taxman2023/k2r.rb]: I've updated the EI constants to include QC_EI_MAX for this purpose

File: lib/taxman/taxman2023/k2r.rb (hunk 7 above)
Comment (mattravenhill-humi) [reply lib/taxman/taxman2023/k2r.rb]: Yep, these constants are now referenced by some helper methods that now live in  lib/taxman/taxman2023/k2_generic.rb

**Comment by mattravenhill-humi** [comment]
@claus-humi @mark-h-humi 
I've updated the implementation to separate the K2(K2Q/K2R/K2QR) and K2P calculations. 
This is now handled by lib/taxman/taxman2023/k2_generic.rb and lib/taxman/taxman2023/k2_p_generic.rb.

Along with this change I've defined the newly required constants and added some basic specs.

**Review by claus-humi — COMMENTED** [review]
I think there's some confusion as to the calculation of K2 as K2 (and K2R/K2Q/K2RQ) are for calculating the federal taxes in T3, and K2P is only used for non QC provinces. It seems strange to me that we are calculating K2 from K2P?

File: lib/taxman/taxman2023/tax_calculator.rb [hunk 10]
```diff
@@ -43,7 +43,7 @@ def calculate
       check_required_context
 
       # Prep factors for T3
-      context[:k2] = K2.amount(context)
+      context[:k2] = K2P.amount(context)
```
[the remark this answers is not in the document]
Comment (mattravenhill-humi) [reply lib/taxman/taxman2023/tax_calculator.rb]: Updated to use the proper K2 class

File: lib/taxman/taxman2023/calculate.rb [hunk 11]
```diff
@@ -40,6 +40,10 @@ def call
 
       context[:a] = A.amount(context)
 
+      context[:k2q] = K2Q.amount(context)
+      context[:k2r] = K2R.amount(context)
+      context[:k2rq] = K2RQ.amount(context)
+
       # The non bonus tax calculation should not have the bonus, but we need it for the bonus tax calculation
```
[the remark this answers is not in the document]
Comment (mattravenhill-humi) [reply lib/taxman/taxman2023/calculate.rb]: These value calculations have now been moved into the tax_calculator

File: spec/taxman/taxman2023/k2_p_spec.rb [hunk 12]
```diff
@@ -1,8 +1,8 @@
 # frozen_string_literal: true
 
-RSpec.describe Taxman2023::K2P do
```
Comment (claus-humi) [inline spec/taxman/taxman2023/k2_p_spec.rb]: Spec file should be renamed to reflect class being tested.

**Review by claus-humi — COMMENTED** [review]
Almost there it seems :)

File: spec/taxman/taxman2023/k2_p_spec.rb (hunk 12 above)
Comment (mattravenhill-humi) [reply spec/taxman/taxman2023/k2_p_spec.rb]: the spec name has been updated

**Review by claus-humi — APPROVED** [review]

**Pull request merged**